File: autocorrelation.py
import matplotlib.pyplot as plt
import pandas as pd
import re
import seaborn as sns

class Autocorrelation():
    """
    A class for the cryptanalysis of the Vigenère
    cipher used to determine key length.
    """

    def __init__(self, ciphertext: str = None) -> None:
        """
        Initializes an Autocorrelation-object
        with or without a ciphertext
        """
        self.ciphertext = ciphertext
    
    def get_ciphertext(self) -> str:
        """
        Returns object's current ciphertext
        """
        return self.ciphertext
    
    def set_ciphertext(self, text: str) -> None:
        """
        Sets object's current ciphertext
        """
        self.ciphertext = text
    
    def normalize(self) -> None:
        """
        Preprocess input text to upper case,
        remove punctuation, spacing and
        newlines / carriage returns (\r: Windows)
        """
        self.ciphertext = self.ciphertext.upper()

        self.ciphertext = re.sub(r'[^\w\s]', '', self.ciphertext)
        self.ciphertext = self.ciphertext.replace(" ", "").replace("\n", "").replace("\r", "")
    
    def normalize_german(self) -> None:
        """
        Preprocess input text to upper case,
        remove punctuation, spacing and
        newlines / carriage returns (\r: Windows),
        and replace characters unique to the
        German language
        """
        self.normalize()

        self.ciphertext = self.ciphertext.replace("Ä", "AE")
        self.ciphertext = self.ciphertext.replace("Ö", "OE")
        self.ciphertext = self.ciphertext.replace("Ü", "UE")
        self.ciphertext = self.ciphertext.replace("ß", "SS")
    
    def shift_text(self, positions: int) -> str:
        """
        Shifts a copy of ciphertext an 
        input number of positions
        """
        return (positions * "-") + self.ciphertext[0:len(self.ciphertext) - positions]
    
    def print_shifts(self) -> None:
        """
        A helper function which iterates over
        and prints all shifts of text
        """
        for i in range(len(self.ciphertext)):
            print(i, ": ", self.shift_text(i))

    def compare_texts(self, shifted_text: str) -> int:
        """
        Compares ciphertext and shifted
        text and returns the number of
        matching letters for each index
        """
        matches = 0

        for i in range(len(shifted_text)):
            if(self.ciphertext[i] == shifted_text[i]):
                matches += 1
        
        return matches
    
    def autocorrelate1(self, number_shifts: int) -> None:
        """
        Prints the number of matching letters
        resulting from a comparison of the
        ciphertext and each shift of a copy
        of the ciphertext up to a supplied
        number of shifts
        """
        for i in range(number_shifts):

            st = self.shift_text(i)

            count = self.compare_texts(st)

            print(i, ": ", count)

    def autocorrelate2(self, number_shifts: int) -> None:
        """
        Returns as a dictionary the number of
        matching letters resulting from a
        comparison of the ciphertext and each
        shift of a copy of the ciphertext up
        to a supplied number of shifts
        """
        matches = {}

        for i in range(number_shifts):
            
            st = self.shift_text(i)

            matches[i] = self.compare_texts(st)
        
        self.display_correlations(matches)
    
    def display_correlations(self, dictionary: dict) -> None:
        """
        A helper function to display
        matches per shift as a line chart
        """
        dictionary.pop(0) # remove first key-value pair to avoid skewing graph

        plt.figure(figsize=(12, 6))
        df = pd.Series(dictionary)
        new_sns = sns.lineplot(data = df, marker="o")
        new_sns.set(xticks=df.keys())

        plt.title('Autocorrelation analysis')
        plt.xlabel('Shifts')
        plt.ylabel('Matches')

        plt.grid()
        plt.show()

ac = Autocorrelation("äö..üü")
ac.normalize_german()
print(ac.get_ciphertext())

File: Autocorrelation.java
import java.util.HashMap;
import java.util.Map;

public class Autocorrelation {

    private String cipherText;

    /*
     * Allows for the initialization of an
     * Autocorrelation-object without a
     * ciphertext.
     */
    public Autocorrelation() { }
    
    /*
     * Allows for the initialization of an
     * Autocorrelation-object with a ciphertext.
     */
    public Autocorrelation(String cipherText) {
        this.cipherText = cipherText;
    }

    /*
     * Returns object's current ciphertext.
     */
    public String getCipherText() {
        return this.cipherText;
    }

    /*
     * Sets object's current ciphertext.
     */
    public void setCipherText(String text) {
        this.cipherText = text;
    }

    /*
     * Preprocess input text to upper case,
     * remove punctuation, spacing and
     * newlines / carriage returns.
     */
    public void normalize() {
        this.cipherText = this.cipherText.toUpperCase();
        this.cipherText = this.cipherText.replaceAll("[\\p{Punct}\\s]+", "");
    }

    /*
     * Preprocess input text to upper case,
     * remove punctuation, spacing and
     * newlines / carriage returns,
     * and replace characters unique to the
     * German language.
     */
    public void normalizeGerman() {
        this.cipherText = this.cipherText.toUpperCase();
        this.cipherText = this.cipherText.replaceAll("[\\p{Punct}\\s]+", "");

        this.cipherText = this.cipherText.replace("Ä", "AE");
        this.cipherText = this.cipherText.replace("Ö", "OE");
        this.cipherText = this.cipherText.replace("Ü", "UE");
        this.cipherText = this.cipherText.replace("ß", "SS");
    }

    /*
     * Shifts a copy of ciphertext an
     * input number of positions.
     */
    public String shiftText(int positions) {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < positions; i++) { sb.append("-"); }
        sb.append(this.cipherText.substring(0, this.cipherText.length() - positions));
        return sb.toString();
    }

    /*
     * A helper function which iterates over
     * and prints all shifts of text.
     */
    public void printShifts() {
        for(int i = 0; i < this.cipherText.length(); i++) {
            System.out.println((i < 10 ? "0" + i : i) + ":\t" + this.shiftText(i));
        }
    }

    /*
     * Compares ciphertext and shifted
     * text and returns the number of
     * matching letters for each index.
     */
    public int compareTexts(String shiftedText) {
        int matches = 0;
        for(int i = 0; i < shiftedText.length(); i++) {
            if(this.cipherText.charAt(i) == shiftedText.charAt(i)) { matches++; }
        }
        return matches;
    }

    /*
     * Prints the number of matching letters
     * between ciphertext and each shift of
     * the copy of the ciphertext up to a
     * supplied number of shifts.
     */
    public void autocorrelate1(int shifts) {
        int i = 0;
        while(i < shifts) {
            String st = this.shiftText(i);
            int count = this.compareTexts(st);
            System.out.println(i++ + ": " + count);
        }
    }

    /*
     * Returns a map containing the number
     * of matching letters between ciphertext and
     * each shift of the copy of the ciphertext
     * up to a supplied number of shifts.
     */
    public Map<Integer, Integer> autocorrelate2(int shifts) {
        Map<Integer, Integer> matches = new HashMap<>();
        int i = 0;
        while(i < shifts) {
            String st = this.shiftText(i);
            int count = this.compareTexts(st);
            matches.put(i++, count);
        }
        matches.remove(0); // remove first key-value pair to avoid skewing graph
        return matches;
    }
}